src/cvc.ml

```diff
@@ -0,0 +1,280 @@
+        then begin
+          let i = i + 1 in
+          (tokenize[@tailcall]) ~acc:((`Length code) :: acc) bytes i lim state (Value code)
+        end
+        else
+          begin match code with
+            | 0x81 ->
+                let code = Cstruct.get_uint8 bytes (i + 1) in
+                let i = i + 2 in
+                (tokenize[@tailcall]) ~acc:((`Length code) :: acc) bytes i lim state (Value code)
+            | 0x82 ->
+                let code = Cstruct.BE.get_uint16 bytes (i + 1) in
+                let i = i + 3 in
+                (tokenize[@tailcall]) ~acc:((`Length code) :: acc) bytes i lim state (Value code)
+            | _ ->
+                raise (Failure "Invalid LENGTH field in TLV encoded CVC data")
+          end
+    | Value length ->
+        let is_rec =
+          match state with
+            | None -> false
+            | Some (_, (_, x)) -> x
+        in
+        let acc =
+          if is_rec
+          then
+            `Value (tokenize ~acc:[] bytes i (i + length) None Init) :: acc
+          else
+            let bytes' =
+              Cstruct.sub bytes i length
+            in
+            `Bytes bytes' :: acc
+        in
+        (if length + i >= Cstruct.len bytes then List.rev acc else (tokenize[@tailcall]) ~acc bytes (i + length) lim None Init)
+  in
+  let tokens = tokenize ~acc:[] bytes 0 (Cstruct.len bytes) None Init in
+  let rec parse = function
+    | `Type (_, (`PUBLIC_KEY, _)) :: `Length _ :: `Value ls :: _ ->
+        parse ls
+    | `Type (_, (`OID, _)) :: `Length _ :: `Bytes bytes :: tl ->
+        let bytes =
+          let prefix =
+            Printf.sprintf "\006%c" (Char.chr (Cstruct.len bytes))
+            |> Cstruct.of_string
+          in
+          Cstruct.append prefix bytes
+        in
+        `Oid (decode_oid bytes) :: parse tl
+    | `Type (_, (`MODULUS, _)) :: `Length _ :: `Bytes bytes :: tl ->
+        `Modulus (atoz_bigendian bytes) :: parse tl
+    | `Type (0x82, ((*`EXPONENT*) _ , _)) :: `Length _ :: `Bytes bytes :: tl ->
+        `Exponent (atoz_bigendian bytes) :: parse tl
+    | `Type (_, (`COEFFICIENT_B, _)) :: `Length _ :: `Bytes bytes :: tl ->
+        `Coefficient_b (atoz_bigendian bytes) :: parse tl
+    | `Type (_, (`BASE_POINT_G, _)) :: `Length _ :: `Bytes bytes :: tl ->
+        `Base_point_g (atoz_bigendian bytes) :: parse tl
+    | `Type (_, (`BASE_POINT_R_ORDER, _)) :: `Length _ :: `Bytes bytes :: tl ->
+        `Base_point_r_order (atoz_bigendian bytes) :: parse tl
+    | `Type (_, (`PUBLIC_POINT_Y, _)) :: `Length _ :: `Bytes bytes :: tl ->
+        `Public_point_y (atoz_bigendian bytes) :: parse tl
+    | `Type (_, (`COFACTOR_F, _)) :: `Length _ :: `Bytes bytes :: tl ->
+        `Cofactor_f (atoz_bigendian bytes) :: parse tl
+    | [] ->
+        []
+    | `Type (t, _) :: tl ->
+        parse tl
+    | `Length _ :: tl ->
+        parse tl
+    | `Bytes _ :: tl ->
+        parse tl
+    | `Value _ :: tl ->
+        parse tl
+    | _ ->
```

Comment:
Changed, turns out that the few cases before the wildcard will catch everything we care about, but the compiler wasn't giving an "unused match case" warning because of the behaviour of inferred polymorphic variants with the wildcard.